src/contexts/NetworkInfo.js

```diff
@@ -42,24 +62,81 @@ export default function Provider({ children }) {
     })
   }, [])
 
+  const updateTokensList = useCallback((chainId, tokensList) => {
+    dispatch({
+      type: UPDATE_TOKENS_LIST,
+      payload: {
+        chainId,
+        tokensList,
+      },
+    })
+  }, [])
+
   return (
     <NetworkInfoContext.Provider
       value={useMemo(
         () => [
           state,
           {
             updateChain,
+            updateTokensList,
           },
         ],
-        [state, updateChain]
+        [state, updateChain, updateTokensList]
       )}
     >
       {children}
     </NetworkInfoContext.Provider>
   )
 }
 
+async function getTokenList(listUrl) {
+  let response
+  try {
+    response = await fetch(listUrl)
+  } catch (error) {
+    console.debug('Failed to fetch list', listUrl, error)
+    throw new Error(`Failed to download list ${listUrl}`)
+  }
+
+  if (!response.ok) {
+    throw new Error(`Failed to download list ${listUrl}`)
+  }
+
+  const json = await response.json()
+  // if (!tokenListValidator(json)) {
+  //   const validationErrors: string =
+  //     tokenListValidator.errors?.reduce<string>((memo, error) => {
+  //       const add = `${error.dataPath} ${error.message ?? ''}`
+  //       return memo.length > 0 ? `${memo}; ${add}` : `${add}`
+  //     }, '') ?? 'unknown error'
+  //   throw new Error(`Token list failed validation: ${validationErrors}`)
+  // }
+  return json
+}
+
 export function useNetworksInfo() {
   const [{ networksInfo }, { updateChain }] = useNetworksInfoContext()
   return [networksInfo.filter(Boolean).length ? networksInfo : [NetworksInfoEnv[0]], updateChain]
 }
+
+export function useTokensList() {
+  const [networksInfo] = useNetworksInfo()
+  const [{ tokensList }, { updateTokensList }] = useNetworksInfoContext()
+
+  useEffect(() => {
+    const fetchTokensList = async index => {
+      try {
+        const tokenList = await getTokenList(networksInfo[index].tokensListUrl)
+        updateTokensList(networksInfo[index].chainId, tokenList)
+      } catch (e) {}
+    }
+    networksInfo.forEach((networkInfo, index) => {
+      if (!tokensList?.[networkInfo.chainId]) {
+        memoRequest(() => fetchTokensList(index), 'useTokensList_' + networkInfo.chainId, 9999999999) //never expired cache
```

Comment:
I see that this is only cache on memory, do we need this? 
for a better cache, I suggest using service worker. You can research about it but for now, I think we dont need cache here

Reply:
Because this is workaround for React context. Context dont have optimize reload like redux, it reload all subcriber everytime it change so these api call would call multiple times at the same time when it other states change. So I put the memo here to cache the first call, the folowing would take the first call without needs to call again.

Of couse, In this specific case, we would not need to use memo here because currently we only have 2 state in this context, so I apply it better than not for future we may add some more states.